lib/zzhc/zzhc.c

```diff
@@ -0,0 +1,600 @@
+	}
+
+	memcpy(ctx->ueiccid, cptr, ICCID_LEN);
+	return rc;
+}
+
+/**@brief Finite state machine */
+static int fsm(struct zzhc *ctx)
+{
+	int rc;
+
+	switch (ctx->state) {
+	case STATE_INIT:
+		LOG_DBG("STATE_INIT");
+
+		sem_init(ctx->sem, 0, 0);
+		register_handler(ctx, at_notif_handler);
+		ctx->state = STATE_WAIT_FOR_REG;
+		break;
+
+	case STATE_WAIT_FOR_REG:
+		LOG_DBG("STATE_WAIT_FOR_REG");
+
+		sem_wait(ctx->sem);
+		ctx->state = (check_iccid(ctx) == 0) ? STATE_END : STATE_START;
+		break;
+
+	case STATE_START:
+		if (ctx->registered) {
+			/* Compile HTTP request. */
+			build_http_req(ctx);
+
+			/* Send. */
+			if (resolve_and_connect(ctx) != 0) {
+				ctx->state = STATE_RETRY_TIMEOUT;
+				break;
+			}
+
+			if (post_request_send(ctx) != 0) {
+				ctx->state = STATE_RETRY_TIMEOUT;
+				break;
+			}
+
+			ctx->rx_offset = 0;
+			ctx->delimiter = 0;
+			ctx->header_recvd = false;
+			ctx->state = STATE_HTTP_REQ_SENT;
+		}
+		break;
+
+	case STATE_HTTP_REQ_SENT:
+		LOG_DBG("STATE_HTTP_REQ_SENT");
+
+		rc = receive_response(ctx);
+		if (rc < 0) {
+			LOG_DBG("Error reading from socket, errno=%d", errno);
+			ctx->state = STATE_RETRY_TIMEOUT;
+			break;
+		}
+
+		if (rc == 0) {
+			LOG_DBG("Peer closed connection!");
+			ctx->state = STATE_HTTP_RESP_RECVD;
+			break;
+		}
+		break;
+
+	case STATE_HTTP_RESP_RECVD:
+		LOG_DBG("STATE_HTTP_RESP_RECVD");
+
+		/* Parse HTTP response. */
+		if (check_http_payload(ctx)) {
+			/* Try to write ICCID to local storage. */
+			rc = save_iccid(ctx->ueiccid, ICCID_LEN);
+			if (rc <= 0) {
+				LOG_DBG("save_iccid() = %d", rc);
+			}
+
+			ctx->state = STATE_END;
+		} else {
+			ctx->state = STATE_RETRY_TIMEOUT;
+		}
+		break;
+
+	case STATE_RETRY_TIMEOUT:
+		LOG_DBG("STATE_RETRY_TIMEOUT");
+
+		/* Cleanup */
+		disconnect(ctx);
+		ctx->http_pkt_len = 0;
+		memset(ctx->http_pkt, 0, HTTP_PKT_LEN);
+
+		if (ctx->retry_cnt >= RETRY_CNT_MAX) {
+			ctx->state = STATE_END;
+			break;
+		}
+
+		sleep(RETRY_TIMEOUT);
+		ctx->retry_cnt++;
+		ctx->state = (ctx->registered)
+			? STATE_START : STATE_WAIT_FOR_REG;
+		break;
+
+	case STATE_END:
+		LOG_DBG("STATE_END");
+
+		/* Clean-up */
+		disconnect(ctx);
+		deregister_handler(ctx, at_notif_handler);
+		ext_uninit(ctx);
+		this_free(ctx);
+		return -ECANCELED;
+
+	default:
+		LOG_DBG("Unknown state=%d", (int)ctx->state);
+		return -ECANCELED;
+	}
+	return 0;
+}
+
+/**@brief Thread function. */
+static void thread_fn(void *arg1, void *arg2, void *arg3)
+{
+	struct zzhc *ctx = (struct zzhc *)arg1;
+
+	while (true) {
+		if (fsm(ctx) == -ECANCELED) {
+			return;
+		}
+	}
+}
+
+/**@brief Function for initializing self-registration daemon. */
+static int zzhc_init(struct device *arg)
+{
+	int rc;
+	struct zzhc *ctx;
+
+	/* Unconditionally subscribe +CEREG events */
+	if (at_cmd_xfer("AT+CEREG=5", NULL, 0) != 0) {
+		LOG_DBG("Can't subscribe to +CEREG events.");
+		return -EIO;
+	}
+
+	/* Allocate resource. */
+	ctx = (struct zzhc *)this_malloc(sizeof(struct zzhc));
```

Comment:
Could we use a static memory instead of heap for context data? Zephyr heap is known for not-handling-well large allocations, so this might cause trouble for smaller heaps. It is under a rework, but who knows when will this be fixed...

Reply:
This is the tricky part.

The zzhc specification says, "After execution, it should quit and not be persistent." Therefore, the memory need to be freed. With `k_free()`, we need `k_malloc()` 

Besides, using this kind of malloc-and-free approach (i.e. new and delete object) allows for higher flexibility in the future so that multiple instances of the library can be instantiated. With slight modifications in the future, we can launch multiple threads to connect to different servers.

The allocated memory fragments are not really large:
- struct zzhc has 644 bytes. (This will become 640 bytes in the next push.)
- struct k_thread has 176 bytes. (This will become static in the next push.)
- struct k_sem has 24 bytes.
- struct at_param_list has 8 bytes.
- struct settings_handler has 24 bytes.

Will these make Zephyr fail?

Reply:
Not exactly fail, but even all these allocations are < 1 kB, they would require ~3kB heap size with current Zephyr heap implementation.

Reply:
`thread `has been removed, whereas `storage `has become static variable global to the file.
While the zzhc_port.c is still needed at the moment, `sem` and `at_list` still need to be allocated. They will become part of `struct zzhc` once zzhc_port.c is merged to zzhc.c.